src/components/dragZoomLayer.ts

```diff
@@ -0,0 +1,152 @@
+     * Usage: Construct the selection box layer and attach x and y scales, and then add the layer
+     * over the plot you are zooming on using a Component Group.
+     */
+    constructor(xScale: QuantitativeScale<Numeric>,
+                yScale: QuantitativeScale<Numeric>) {
+      super();
+      this.xScale(xScale);
+      this.yScale(yScale);
+      this._dragInteraction = new Interactions.Drag();
+      this._dragInteraction.attachTo(this);
+      this._doubleClickInteraction = new Interactions.DoubleClick();
+      this._doubleClickInteraction.attachTo(this);
+      this.setupCallbacks();
+    }
+
+    private setupCallbacks() {
+      let dragging = false;
+      this._dragInteraction.onDragStart((startPoint: Point) => {
+        this.bounds({
+          topLeft: startPoint,
+          bottomRight: startPoint
+        });
+      });
+      this._dragInteraction.onDrag((startPoint, endPoint) => {
+        this.bounds({topLeft: startPoint, bottomRight: endPoint});
+        this.boxVisible(true);
+        dragging = true;
+      });
+      this._dragInteraction.onDragEnd((startPoint, endPoint) => {
+        this.boxVisible(false);
+        this.bounds({topLeft: startPoint, bottomRight: endPoint});
+        if (dragging) {
+          this.zoom();
+        }
+        dragging = false;
+      });
+
+      this._doubleClickInteraction.onDoubleClick(() => this.unzoom());
+    }
+
+    /* Set the time (in ms) over which the zoom will interpolate.
+     * 0 implies no interpolation. (ie zoom is instant)
+     */
+    public animationTime(): number;
+    public animationTime(animationTime: number): DragZoomLayer;
+    public animationTime(animationTime?: number): any {
+      if (animationTime == null) {
+        return this._animationTime;
+      }
+      if (animationTime < 0) {
+        throw new Error("animationTime cannot be negative");
+      }
+      this._animationTime = animationTime;
+      return this;
+    }
+
+    /* Set the easing function, which determines how the zoom interpolates over time. */
+    public ease(fn: (t: number) => number): DragZoomLayer {
+      if (typeof(fn) !== "function") {
+        throw new Error("ease function must be a function");
+      }
+      if (fn(0) !== 0 || fn(1) !== 1) {
+        Utils.Window.warn("Easing function does not maintain invariant f(0)==0 && f(1)==1. Bad behavior may result.");
+      }
+      this.easeFn = fn;
+      return this;
+    }
+
+    // Zoom into extent of the selection box bounds
+    private zoom() {
+      let x0: number = this.xExtent()[0].valueOf();
+      let x1: number = this.xExtent()[1].valueOf();
+      let y0: number = this.yExtent()[1].valueOf();
+      let y1: number = this.yExtent()[0].valueOf();
+
+      if (x0 === x1 || y0 === y1) {
+        return;
+      }
+
+      if (!this.isZoomed) {
+        this.isZoomed = true;
+        this.xDomainToRestore = this.xScale().domain();
+        this.yDomainToRestore = this.yScale().domain();
```

Comment:
Is it possible the domain will change after the scale is zoomed? These may become out of date